Identity provider should not be hidden from login page when the a user goes back from the reset credentials page

vilmosnagy opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->
Closes #33204.

Breaks only one test, `RegisterTest::testRegisterShouldFailBeforeUserCreationWhenUserIsInContext`, which was added in #23064 (issue #21514) - that test covers basically the same bug (also reported by me). In the previous bug the steps to reproduce was:
 - go to reset credentials, and pick a valid username
 - then go back, and try to register -> the fix was to check and raise an error during the registration if the user was set in the `ResetCredentialChooseUser`

The current issue (#33204) I'm having in short is:
 - go to reset credentials, and pick a valid username
 - then go back to the login page -> the social logins, which doesn't linked to the picked account, will be hidden from the login. You don't even need to pass the authentication to find out what social sites the user has linked to their account.

My current fix checks whether a user has been set in the context during an attempted (and cancelled) reset credential flow - and if it has been set, then we clear them from the context when loading the login page.

If I modify the failing test (also written by me), would you be willing to merge this?  Probably I need to:
 - modify the failing test
 - add a new, slightly modified `RegisterTest::testRegisterShouldFailBeforeUserCreationWhenUserIsInContext` one where the login page is skipped entirely

cc @ahus1 @sguilhen @mposolda who was involved in the test which fails currently.

vilmosnagy committed:
#33204 add the failing test

Signed-off-by: Vilmos Nagy <vilmos.nagy@outlook.com>

vilmosnagy committed:
#33204 possible fix (?)

Signed-off-by: Vilmos Nagy <vilmos.nagy@outlook.com>

jonkoops commented:
Thanks for the PR! FYI if you want to link an issue you can [use a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to do so. We try to avoid mentioning the issue in the title. I went ahead and edited the issue description to link it up.

sguilhen commented:
@vilmosnagy I haven't had the time to look into your PR yet, but the test failures seem to be related - actually only one test seems to be failing - a registration test.

vilmosnagy committed:
#33204 refactor the previous tests to work with the changed behaviour

Signed-off-by: Vilmos Nagy <vilmos.nagy@outlook.com>

# Please enter the commit message for your changes. Lines starting
# with '#' will be kept; you may remove them yourself if you want to.
# An empty message aborts the commit.
#
# Date:      Sun Oct 6 16:35:19 2024 +0200
#
# On branch issue_33204
# Changes to be committed:
#	modified:   testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/RegisterTest.java
#

vilmosnagy commented:
@sguilhen thanks for your feedback, I've pushed the changes I described. As far as I see the failing tests are not connected to this PR (they seem to be js & db related, while I only touched registration-flow related code).

Could you please move forward with the review? Thanks

sguilhen commented:
@ahus1 Could you please re-run the CI for this PR? Failures don´t seem related to the changes

sguilhen commented:
@vilmosnagy @ahus1 Yeah, this test has been failing in other PRs, but IIRC it was fixed last week (see https://github.com/keycloak/keycloak/issues/33440). Have you rebased your branch recently?

ahus1 commented:
Rebased just now, let's see what the tests report.

vilmosnagy commented:
@sguilhen as far as I remember I've rebased it during the weekend. But it fails even after @ahus1 's rebase as well :'(

sguilhen commented:
@edewit I thought this duplicate group test was disabled last week, but it seems it is still happening like in this PR. Any ideas what we could do?

edewit commented:
@sguilhen I thought it was skipped as well, it seems it's not the case I've added a commit that disables this test to this PR

vilmosnagy commented:
Is that possible that the failing `test03SpnegoLoginWithCorrectKerberosPrincipalRealm` test is a flaky one? It was detected by the bot in previous issues with a very similar stacktrace & error: https://github.com/keycloak/keycloak/pull/32283#issuecomment-2313090758

sguilhen commented:
> Is that possible that the failing `test03SpnegoLoginWithCorrectKerberosPrincipalRealm` test is a flaky one? It was detected by the bot in previous issues with a very similar stacktrace & error: [#32283 (comment)](https://github.com/keycloak/keycloak/pull/32283#issuecomment-2313090758)

Yeah, it is likely a flaky one that needs to be looked into

mposolda reviewed (commented):
@vilmosnagy Thanks for the PR! I think that the change in the test `RegisterTest.testRegisterShouldFailBeforeUserCreationWhenUserIsInContext` is OK IMO. 

However I have one more scenario I would like to discuss, when user opens link from the email after click "Back". In details:

1) User clicks "Forget password" on the login screen
2) User chooses the username `john` (Now `john` useris set in the authenticationSession) and there is email sent to the email address of `john`
3) Now user click "Back" and login screen is displayed
4) Now user goes to his email and clicks on the link from the email from step (2). Now what happens?

It seems to me that before this PR, user is verified and he is able to reset his password and then he is authenticated.

But can you check what happen after your changes?

I did not check, but I think that after your changes, the user is not verified as he is not anymore in the context of authenticationSession. IMO this change is fine if there is some sensible page displayed to the user like `Action expired. Please continue here` with the link where user can restart his login and continue from the start. But what is not OK is some unexpected error page like NullPointerException (I did not tried, but I have some suspicion that NPE can happen in `AbstractActionTokenHandler.verifyEmail` due the user cleared from auth session, but maybe I am wrong :-)

vilmosnagy commented:
Hi @mposolda,

Thanks for your review.

> But can you check what happens after your changes?

You're probably right—my changes might break the forgot password flow if the user navigates back to the login page before clicking on the email link.

One question, though: do we need to show an action expired page in this case? For example, if I start the password reset flow on one device (e.g., mobile) and click the reset password link on another (e.g., desktop), the reset is allowed in the new session. The only difference is:

 - If the user opens the link in the original session, they are redirected back to the app in a logged-in state.
 - If the user opens the link in a new session, an informational page is displayed with a prompt to return to the base URL of the client.

Would this approach be feasible for you? Just theorizing here, but maybe something like:

1. The user clicks "Forgot password" on the login screen.
2. The user enters the username john (now john is set in the authentication session), and an email is sent to john's email address.
3. The user clicks "Back," and the login screen is displayed.
4. The user clicks the link in the email from step 2.
5. John is allowed to set the new password and is redirected to client.baseUrl upon success.

Either way: I see your points here, and I'll need a couple of days to write a few more tests. I'll get back to you once those are done.

Thanks again,
Vilmos

mposolda commented:
@vilmosnagy Yes, I think you are right. It will be better user experience if the user is authenticated similarly like in the current Keycloak main instead of displaying `Action expired` page.

The only thing is, in the described scenario, the user might not be anymore in the context of authenticationSession after your changes. So perhaps in that case, it may be OK to add user back to the authenticationSession based on the info from the actionToken? I can see that `VerifyEmailActionToken` has reference to the user, so hopefully it should be possible to restore the user in the authenticationSession.

Anyway, it will be good to add automated test for this scenario and make sure that it works as expected.

Thanks a lot for the discussion and for the additional work on handling those cases!

mposolda commented:
@vilmosnagy Any chance to continue on this PR?

thomasdarimont commented:
@vilmosnagy congratulations :)

vilmosnagy commented:
@mposolda reopened in #48814, and I addressed your thoughts from this review session.
